File: vision/src/third.py
import cv2
import numpy as np
import copy

SCALE = 1
H_SIZE = 512.0

def get_color():
	return np.random.randint(0,255,3)

def get_max_contour(contours):
	idx = -1
	max_val = 0
	for i in range(len(contours)):
		if len(contours[i]) > max_val:
			max_val = len(contours[i])
			idx = i
	return idx

def sortSquare(square):

	# print square
	sorted_square = square[:]
	avg_x, avg_y = np.average(square, axis = 0)

	for s in square:
		if s[0] < avg_x and s[1] < avg_y:
			sorted_square[0] = s
		if s[0] < avg_x and s[1] > avg_y:
			sorted_square[1] = s
		if s[0] > avg_x and s[1] < avg_y:
			sorted_square[2] = s
		if s[0] > avg_x and s[1] > avg_y:
			sorted_square[3] = s

	return sorted_square


	




def locate_puzzle(img):

	# Load image
	#img = cv2.imread('../img/cifar6.jpg', cv2.CV_LOAD_IMAGE_COLOR)
	img = cv2.resize(img, (0,0), fx=1.0/SCALE, fy=1.0/SCALE )
	img_2 = copy.deepcopy(img)
	# cv2.imshow('img',img)
	# cv2.waitKey(0)
	# GreyScaling & Thresholding
	img_gray = cv2.cvtColor(img, cv2.COLOR_BGR2GRAY)
	ret , img_gray = cv2.threshold(img_gray, 0, 255, (cv2.THRESH_BINARY + cv2.THRESH_OTSU));
	

	# Morphological operations
	morph = cv2.dilate(img_gray, cv2.getStructuringElement(cv2.MORPH_RECT,(11/SCALE,11/SCALE)))
	morph = cv2.erode(morph, cv2.getStructuringElement(cv2.MORPH_RECT,(11/SCALE,11/SCALE)))
	# cv2.imshow('morph',morph)
	# cv2.waitKey(0)

	# Edge detection
	canny = cv2.Canny(morph, 100, 200);
	# cv2.imshow('canny',canny)
	# cv2.waitKey(0)
	
	# Contours detection
	contours, i = cv2.findContours(canny, cv2.RETR_EXTERNAL, cv2.CHAIN_APPROX_SIMPLE)

	# Get & Draw Largest Contour
	idx = get_max_contour(contours)
	drawing = np.zeros(np.array(img).shape, dtype=np.uint8)
	cv2.drawContours(img, contours, idx, (0,0,255), 1)
	# cv2.imshow('draw',drawing)
	# cv2.waitKey(0)

	# Approximate Poly DP
	square = []
	approx = cv2.approxPolyDP(contours[idx], cv2.arcLength(contours[idx], True)*0.02, True)
	# print 'Approx size: ', np.array(approx).shape
	point_cnt = 0
	for n in range(len(approx)):
			
		cont = False
		for m in range(n):
			dist = cv2.norm(approx[n]-approx[m])
			#print approx[n], apporx[m], dist
			#raw_input()
			if dist < (30/SCALE):
				cont = True
				break
		if cont == True:
			continue
		else:
			point_cnt += 1
			# if point_cnt > 4:
				# continue
			center = (approx[n][0][0],approx[n][0][1])
			cv2.circle(img, center, 2, (0,0,255), 2)
			square.append( [float(i) for i in center])
			# cv2.imshow('draw',drawing)
			# cv2.waitKey(0)
	if len(square) < 4:
		return None, False

	# print "Before: ", square
	square = sortSquare(square[:4])
	# print "Sort: ", square

	cv2.imshow('img',img)
	# cv2.waitKey(0)
	# Find Homography for Appoximated PolyDP
	transformed = np.zeros((H_SIZE, H_SIZE), dtype=np.uint8)
	dst = [(0.0,0.0),(0.0,H_SIZE-1),(H_SIZE-1,0.0),(H_SIZE-1,H_SIZE-1)]
	h,_ = cv2.findHomography(np.array(square), np.array(dst), 0)
	#print h

	# Warp Perspective Transformation	
	warp = cv2.warpPerspective(img_2, h, transformed.shape)
	cv2.imshow('warp',warp)
	# cv2.waitKey(0)
	return warp, True
	

def crop_cifar_image(img):

	img2 = copy.copy(img)
	# cv2.imshow('img2',img2)
	# cv2.waitKey(0)
	crops = []

	for i in range(16):
		for j in range(16):
			cv2.rectangle(img2, (32*i,32*j), (32*(i+1),32*(j+1)), (0,0,255), 2)
			crops.append(img[32*i:32*(i+1),32*j:32*(j+1)])
	
	cv2.imshow('warp',img2)
	cv2.waitKey(0)

	return crops

if __name__ == '__main__':
	
	cap = cv2.VideoCapture('../img/cifar.mp4')
	cv2.namedWindow('img', cv2.CV_WINDOW_AUTOSIZE)
	cv2.namedWindow('warp', cv2.CV_WINDOW_AUTOSIZE)
	cv2.moveWindow("img", 10, 100);
	cv2.moveWindow("warp", 330, 100);

	while True:
		valid, img = cap.read()

		if not valid:
			break

		cv2.imshow('img', img)
		puzzle, valid = locate_puzzle(img)
		imgs = crop_cifar_image(puzzle)
		cv2.waitKey(0)
		# key = cv2.waitKey(50)
		# if key == 27:
			# break



	# img = cv2.imread('../img/cifar5.jpg', cv2.CV_LOAD_IMAGE_COLOR)
	# puzzle = locate_puzzle(img)
	# imgs = crop_cifar_image(puzzle)

File: vision/src/third.cpp
#include <iostream>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/opencv.hpp>

using namespace cv;
using namespace std;

int lo_diff = 1;
int up_diff = 5;
int th =170; 
int mode = 2;
Mat img, img_2,img_3,img_4, img_hsv,img_grey,img_grey2, img_thresh,img_canny;
Rect ccomp;
void on_trackbar(){
	
	floodFill(img, Point(1,1), Scalar(155, 255,55), &ccomp, Scalar(lo_diff, lo_diff, lo_diff),Scalar(up_diff, up_diff, up_diff)); 
	imshow("fill", img);
}

int main( int argc, char** argv )
{
    // if( argc != 2)
    // {
    //  cout <<" Usage: display_image ImageToLoadAndDisplay" << endl;
    //  return -1;
    // }

    
    img = imread("../img/cifar2.jpg", CV_LOAD_IMAGE_COLOR);   // Read the file
    cout << img.cols/4 <<" , "<<img.rows/4<<endl;
    cv::resize( img, img, cv::Size(img.cols / 4, img.rows / 4) );
    if(! img.data ){
        cout <<  "Could not open or find the image" << std::endl ;
        return -1;
    }
    
    
    //imshow("morph",img);
    cvtColor( img, img_grey, CV_BGR2GRAY );
    threshold(img_grey, img_grey, 0, 255, cv::THRESH_BINARY + cv::THRESH_OTSU);
    //imshow("grey",img_grey);

    
    dilate(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(15,15)));
    erode(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(15,15)));
    //imshow("morph grey1",img_grey);    

    //dilate(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(15,15)));
    //erode(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(15,15)));
    //dilate(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(3,3)) );
    //erode(img_grey, img_grey, getStructuringElement(MORPH_RECT,Size(3,3)));
    
	//imshow("morph grey2",img_grey);    
	//waitKey(0);

	//threshold(img_grey, img_grey2, 1, 255, CV_THRESH_BINARY);
    Canny(img_grey, img_grey2, 100, 200, 3);
    //imshow("grey2",img_grey2);

    vector<vector<Point> > contours;
    vector<Vec4i> hierarchy;
    RNG rng(12345);
    findContours(img_grey2, contours, CV_RETR_EXTERNAL, CV_CHAIN_APPROX_SIMPLE);
    Mat drawing = Mat::zeros( img_grey.size(), CV_8UC3 );

    int idx = -1;
    int max = 0;
    vector<Point> approx;
    vector<Point> square;
    Scalar color;
    for( int i = 0; i< contours.size(); i++ )
    {	
    	//get largest contour
    	if (contours[i].size() > max){
    		max = contours[i].size();
    		idx = i;
    	}
    } 
    drawContours( drawing, contours, idx, Scalar( rng.uniform(0, 255), rng.uniform(0,255), rng.uniform(0,255)));
    approxPolyDP(Mat(contours[idx]), approx, arcLength(Mat(contours[idx]), true)*0.02, true);
    cout<<"Approx size:"<< approx.size()<<endl;
    for (int n=0; n<approx.size(); n++){
    	bool cont = false;
    	for(int m=0; m<n; m++){
    		double dist = cv::norm(approx[n]-approx[m]);
    		if (dist < 10){ //ignore points that are close with eachother
    			cont = true;
    			break;
    		}
    	}
    	if(cont)
    		continue;
    	else{
    		circle(drawing, approx[n], 2, Scalar(rng.uniform(0, 255), rng.uniform(0,255), rng.uniform(0,255)), 2 );
    		square.push_back(approx[n]);		
    		//imshow("draw",drawing);
    		//waitKey(0);
    	}
    		
    }
    
   

   	/*Draw boundary*/
    // int n = (int)square.size();
    // const Point* p = &square[0];
    // polylines(img, &p, &n, 1, true, Scalar(0, 0, 255), 3, 0);
    // imshow("result",img);
    // waitKey(0);


    vector<Point2f> square2f;
    for(int n=0; n<square.size(); n++){
    	float x = (float)square[n].x;
    	float y = (float)square[n].y;
    	cout << x <<" , " <<y<<endl;
    	square2f.push_back(Point2f(x,y));
    }

    Mat tranformed(Size(512,512),CV_8UC3);

    vector<Point2f> dst;
    dst.push_back(Point2f(0, 0));
    dst.push_back(Point2f(511, 0));
    dst.push_back(Point2f(511, 511));
    dst.push_back(Point2f(0, 511));
    

    Mat h = findHomography(square2f, dst, 0);
    cout <<"H"<<h<<endl;
    Mat im_out;
    // Warp source image to destination based on homography
    warpPerspective(img, im_out, h, tranformed.size());
 	
    // Display images
    imshow("Source Image", img);
    //waitKey(0);
    //imshow("Destination Image", tranformed);
    imshow("Warped Source Image", im_out);
    waitKey(0);
    Mat im_out_2 = im_out.clone();
    vector<Rect> crops;
    for(int i=0; i<16; i++){
    	for(int j=0; j<16; j++){
    		rectangle(im_out_2, Point(32*i,32*j), Point(32*(i+1),32*(j+1)), Scalar(0,0,255), 2);
    		crops.push_back(Rect(32*i,32*j,32,32));			
    	}
    }
    imshow("Warped Source Image", im_out_2);
    waitKey(0);    		

    for(int n=0; n<crops.size(); n++){
    	//Mat src(img);
    	Mat cropped(Size(32,32),CV_8UC3);// = image(crops[n]);
    	cropped = im_out(crops[n]).clone();//.copyTo(cropped);

    	
    	imshow("crop", cropped);
    	waitKey(0);    		
    }






    
    return 0;
}

